### PR TITLE
core / frontend: passthrough host support

### PR DESCRIPTION
**Pull request by mcutalo88**
### Description
Add passthrough host support for frontend static assets. If the deployed version of clutch does not have the requested frontend assets on disk the passthrough host will be used to fetch the correct assets assuming they exist for the configured provider.

In a different PR we would have a more pretty failure mode, currently you simply get a white page.

### Testing Performed

Testing was performed internally at Lyft, forcing all static asset traffic to go through the passthrough host.
```
gateway:
  assets:
    s3:
      region: us-east-1
      bucket: clutch
      key: static
```
### GitHub Issue 
Passthrough host support #471

### CODE REVIEW

File: api/config/gateway/v1/gateway.proto [hunk 1]
```diff
@@ -85,6 +85,12 @@ message GatewayOptions {
   Timeouts timeouts = 5;
 
   repeated Middleware middleware = 6;
+
+  Frontend frontend = 7;
+}
+
+message Frontend {
```
Comment (danielhochman) [inline api/config/gateway/v1/gateway.proto:92]:
```suggestion
  Assets assets = 7;
}

message Assets {
```

File: api/config/gateway/v1/gateway.proto [hunk 2]
```diff
@@ -85,6 +85,12 @@ message GatewayOptions {
...
+  string cdn_url = 1;
```
Comment (danielhochman) [inline api/config/gateway/v1/gateway.proto:93]: it's not really a CDN, it's a passthrough. a CDN would go in front of the server whereas this is behind it.

also in our case we use s3, not a publicly accessible URL, so the config will need to a be bit different.

**Review by danielhochman — COMMENTED** [review]
thanks for picking this up. some initial comments on naming clarity.

File: api/config/gateway/v1/gateway.proto (hunk 1 above)
Comment (mcutalo88) [reply api/config/gateway/v1/gateway.proto:92]: Yeah sure sounds good

File: api/config/gateway/v1/gateway.proto (hunk 2 above)
Comment (mcutalo88) [reply api/config/gateway/v1/gateway.proto:93]: Yup makes sense I can change it up. 

Why will the config be different for us? Our IAM role will have access to these assets?

Comment (danielhochman) [reply api/config/gateway/v1/gateway.proto:93]: we need to use an s3 client in order to sign the requests, a regular HTTP request won't do

Comment (mcutalo88) [reply api/config/gateway/v1/gateway.proto:93]: Yeah i see what your saying now after reading a bit, let me compensate.

File: backend/gateway/mux/mux.go [hunk 3]
```diff
@@ -62,14 +69,51 @@ func (a *assetHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 
 	// Serve!
 	if f, err := a.fileSystem.Open(r.URL.Path); err != nil {
+		// If not a known static asset and a asset provider is configured, try streaming from the configured provider.
+		if a.assetCfg.Provider != nil && strings.HasPrefix(r.URL.Path, "/static") {
+			// We attatch this header simply for observibility purposes.
+			// Otherwise its difficult to know if the assets are being served from the configured provider.
+			w.Header().Set("x-clutch-asset-passthrough", "true")
+
+			// TODO: handle errors
```
Comment (mcutalo88) [inline backend/gateway/mux/mux.go:78]: I still need to handle errors some how, I haven't decided on what way would be best yet.

File: api/config/gateway/v1/gateway.proto [hunk 4]
```diff
@@ -85,6 +85,23 @@ message GatewayOptions {
   Timeouts timeouts = 5;
 
   repeated Middleware middleware = 6;
+
+  Assets assets = 7;
+}
+
+// Assets configuration provide a passthrough cdn for frontend static assets.
```
Comment (dschaller) [inline api/config/gateway/v1/gateway.proto:92]:
```suggestion
// Assets configuration provide a passthrough host for frontend static assets.
```

File: backend/gateway/mux/mux.go [hunk 5]
```diff
@@ -62,14 +69,51 @@ func (a *assetHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 
 	// Serve!
 	if f, err := a.fileSystem.Open(r.URL.Path); err != nil {
+		// If not a known static asset and a asset provider is configured, try streaming from the configured provider.
```
Comment (dschaller) [inline backend/gateway/mux/mux.go:72]:
```suggestion
		// If not a known static asset and an asset provider is configured, try streaming from the configured provider.
```

File: backend/gateway/mux/mux.go (hunk 3 above)
Comment (dschaller) [reply backend/gateway/mux/mux.go:78]: We can open another PR this week to handle the error state when either both the host and the pass through fail or only the host if the pass through isn't configured.

File: backend/clutch-config.yaml [hunk 6]
```diff
@@ -15,6 +15,12 @@ gateway:
       address: 0.0.0.0
       port: 8080
       secure: false
+  assets:
+    s3:
```
Comment (dschaller) [inline backend/clutch-config.yaml:19]: Should we have this in the default config? We aren't hosting this bucket so it might cause confusion.

Comment (mcutalo88) [reply backend/clutch-config.yaml:19]: Ah no sorry let me delete this, this was leftover from testing.

**Review by dschaller — APPROVED** [review]

**Pull request merged**